lib: os: spsc_pbuf: Improve data cache configuration

nordic-krch opened the pull request:
Use CONFIG_DCACHE_LINE_SIZE to determine distance between
fields modified by different cores.
Add option which specifies what is the data cache line
of the remote core. Maximum from local and remote cache
line sizes is used as distance and alignement.

This was discussed here https://github.com/zephyrproject-rtos/zephyr/pull/46006#discussion_r910148798 but PR was merged before change was applied.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

carlocaione commented on include/zephyr/sys/spsc_pbuf.h:
```diff
@@ -31,11 +31,19 @@ extern "C" {
 #define SPSC_PBUF_UTILIZATION_OFFSET 8
 
 /**@} */
+#ifdef CONFIG_DCACHE_LINE_SIZE
+#define Z_SPSC_PBUF_LOCAL_DCACHE_LINE CONFIG_DCACHE_LINE_SIZE
+#else
+#define Z_SPSC_PBUF_LOCAL_DCACHE_LINE 0
+#endif
```
You probably want to check for `CONFIG_DCACHE_LINE_SIZE_DETECT` as well. See in include/zephyr/cache.h

nordic-krch replied:
for runtime detection it is a bit tricky because i need static value since it is used for array size declaration.
What i did is to mention in spsc_pbuf cache line configuration that if runtime detection of data cache line size is used then this option must be set to max(local, remote).
I've also added support for getting line size from device tree.

carlocaione commented on include/zephyr/sys/spsc_pbuf.h:
```diff
@@ -31,11 +31,25 @@ extern "C" {
 #define SPSC_PBUF_UTILIZATION_OFFSET 8
 
 /**@} */
+#ifdef DCACHE
```
`#ifdef CONFIG_DCACHE`

uhm, we should probably make `DCACHE_LINE_*` depending on `DCACHE` instead of having this additional check.

nordic-krch replied:
@carlocaione I reworked it a bit. Can you take another look?

carlocaione replied:
Well, I took a look and the usage of `CONFIG_DCACHE` is messy (it was introduced by a later and urelated commit). I need to clean that up but before that can you please:
- Remove dependency on `CONFIG_DCACHE` (sorry about that)
- Just use `CONFIG_DCACHE_LINE_SIZE != 0`? (isn't that implying that `CONFIG_DCACHE_LINE_SIZE` is defined?)

nordic-krch replied:
done

carlocaione commented on include/zephyr/sys/spsc_pbuf.h:
```diff
@@ -31,11 +31,25 @@ extern "C" {
 
 /**@} */
+#ifdef DCACHE
+#ifdef CONFIG_DCACHE_LINE_SIZE
```
`CONFIG_DCACHE_LINE_SIZE` is actually an `int` symbol, so this should be `CONFIG_DCACHE_LINE_SIZE != 0` ?

carlocaione commented on include/zephyr/sys/spsc_pbuf.h:
```diff
@@ -31,11 +31,25 @@ extern "C" {
+#ifdef CONFIG_DCACHE_LINE_SIZE
+#define Z_SPSC_PBUF_LOCAL_DCACHE_LINE CONFIG_DCACHE_LINE_SIZE
+#elif !defined(CONFIG_DCACHE_LINE_SIZE_DETECT)
+#define Z_SPSC_PBUF_LOCAL_DCACHE_LINE DT_PROP_OR(CPU, d_cache_line_size, 0)
+#else
+#define Z_SPSC_PBUF_LOCAL_DCACHE_LINE 0
```
You can merge this two cases (if the DT property doesn't exist, it falls back to 0)

nordic-krch replied:
merged

andyross reviewed (dismissed):
Dumb question: is the hardware involved here something specific to the spsc channel/device/mapping/whatever, or is this just a CPU L1 cache?  If the latter, we probably want to attach this kconfig to the global cache API instead (or maybe dtsify it somehow).

No complaints with the code, obviously.

nordic-krch committed:
lib: os: spsc_pbuf: Improve data cache configuration

Use CONFIG_DCACHE_LINE_SIZE to determine distance between
fields modified by different cores.
Add option which specifies what is the data cache line
of the remote core. Maximum from local and remote cache
line sizes is used as distance and alignement.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch committed:
tests: lib: spsc_pbuf: Increase test buffers

Test buffers must be increased to fit test data when cache
is enabled.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

nordic-krch commented:
@carlocaione @dcpleung @andyross can you reapprove. Had to increase test buffers sizes as they where too small for x86 where 32 byte cache line was used.
